airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.airbyte.config.StandardSyncInput;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import io.airbyte.validation.json.JsonValidationException;
+
+public class RecordSchemaValidator {
+
+  public void validateSchema(final AirbyteMessage message, final StandardSyncInput syncInput) throws RecordSchemaValidationException {
+    // the stream this message corresponds to
+    final String messageStream = message.getRecord().getStream();
+    final JsonNode messageData = message.getRecord().getData();
+    final String streamPrefix = syncInput.getPrefix();
+
+    // the stream name and json schema
+    final ConfiguredAirbyteStream matchingAirbyteStream = syncInput.getCatalog().getStreams().stream()
+        .filter(s -> (String.format(streamPrefix + s.getStream().getName().trim()).equals((messageStream.trim())))).findFirst().orElse(null);
```

Comment:
Right now, this logic to read all of the catalog's streams and filter them down in memory happens for every record that is being validated, which could hurt performance when this is done for thousands or millions of records. We also don't expect this `syncInput` to change from one record to the next.

Instead, this `RecordSchemaValidator` class could take in the `StandardSyncInput` in the constructor, and at construction time it could transform the catalog streams into a map of correctly-formatted stream names to the corresponding ConfiguredAirbyteStream object (i.e. apply `streamPrefix + s.getStream().getName().trim()` to each stream to get the key, with the value being the ConfiguredAirbyteStream itself). 

This way, each individual record validation would just need to check if there is a matching key in the map, and retrieve the corresponding value if so, which is a constant-time operation rather than the O(n) filter operation that we have here.

Reply:
I was starting to implement this, but we don't actually have the StandardSyncInput when the RecordSchemaValidator is constructed. It's constructed when the DefaultReplicationWorker is constructed, but the StandardSyncInput is passed into the `run` method. We could move the SyncInput to be passed into the constructor as well - I think in places where we initialize the DefaultReplicationWorker we also already have the SyncInput -- or do you think the recordSchemaValidator should be moved to be passed into the `run` method?

Reply:
I think moving the SyncInput to be passed into the DefaultReplicationWorker constructor makes sense - as you said it looks like we need the sync input at construction time anyway in order to initialize things like the NamespacingMapper, so this doesn't seem like a huge lift to me. Plus, the jobId is already something we pass to that constructor, and the syncInput should always be the same for a given job id, so conceptually the DefaultReplicationWorker is already scoped to a single job with a single syncInput.

Reply:
In fact, it might make sense for us to just have a single constructor for the DefaultReplicationWorker that takes in the RecordSchemaValidator, and that RecordSchemaValidator takes in the SyncInput in its constructor. This way we are only adding one new constructor parameter to DefaultReplicationWorker, since that is all we would be using the SyncInput for anyway.

Reply:
Also one more note - it would probably be slightly more efficient for the map that I described earlier in this comment chain to just store the corresponding stream's `jsonSchema` as the value, rather than storing the entire ConfiguredAirbyteStream as map value, since the jsonSchema is all we really need.